### ISSUE TITLE
Requiring orthogonality of rotation matrix for SO3 construction is annoying

### ISSUE THREAD

**Issue by tanjunyao7** [strasdat/Sophus#365]
Hi,
I noticed the new features in latest release of Sophus so I decided to update my previous version 1.0 to the new one. However, whenever I construct a SE3 object from an Eigen matrix Sophus complains about that the top left 3x3 block is not orthogonal because it's determinant is not 1.0, even though it's 0.999. I was forced to manually orthogonalize the top left 3x3 block which is a little bit annoying. 
I don't understand why should the user instead of the constructor do this orthogonalization job.

**Comment by VladyslavUsenko** [comment]
You can get the previous behaviour by defining `SOPHUS_DISABLE_ENSURES`.

**Comment by strasdat** [comment]
tldr: just call ``SO3::fitToSO3`` instead of ``SO3`` constructor to construct a 3d rotation and you will be fine.

Longer answer:

There are some complexity / tradeoffs here. To construct SO3 from a matrix you have three options:

 1. Fail if the input matrix is not orthogonal (call SO3 constr)
 2. Take in arbitrary matrix and fit it to be orthogonal with det(R)=1 (``SO3::fitToSO3``)
 3. Define SOPHUS_DISABLE_ENSURES and SO3 construction won't panic on preconditions
 
Point 3. is the most dangerous, since you will get wrong / meaningless results if the matrix is not orthogonal or determined is not ~ 1. E.g. for det(R)=0.999, the results would be only slightly wrong, but you need to place the threshold for the numerical check somewhere.

Point 2. is a good default. However it adds some computational complexity - which can be avoided by calling 1.
 -

**Comment by tanjunyao7** [comment]
> tldr: just call `SO3::fitToSO3` instead of `SO3` constructor to construct a 3d rotation and you will be fine.
> 
> Longer answer:
> 
> There are some complexity / tradeoffs here. To construct SO3 from a matrix you have three options:
> 
> 1. Fail if the input matrix is not orthogonal (call SO3 constr)
> 2. Take in arbitrary matrix and fit it to be orthogonal with det(R)=1 (`SO3::fitToSO3`)
> 3. Define SOPHUS_DISABLE_ENSURES and SO3 construction won't panic on preconditions (as @VladyslavUsenko said)
> 
> Point 3. is the most dangerous, since you will get wrong / meaningless results if the matrix is not orthogonal or determined is not ~ 1. E.g. for det(R)=0.999, the results would be only slightly wrong, but you need to place the threshold for the numerical check somewhere.
> 
> Point 2. is a good default. However it adds some computational complexity - which can be avoided by calling 1.

Personally I think point 2 is the best choice. The extra computational complexity in construction is quite negligible because in SO3 or SE3 multiplication, which I believe is the most use case for Sophus, an intentional orthogonalization step after matrix multiplication is always done. Why can’t the constructor follow the same design philosophy?

**Comment by VladyslavUsenko** [comment]
It is not true that extra computational complexity is negligible. If you look at the code (https://github.com/strasdat/Sophus/blob/d1374a9e30d1ec9d7b8295e6f14dfa4d6569337a/sophus/rotation_matrix.hpp#L61-L79), it uses SVD to orthogonalize matrices, which is computationally expensive. So from my opinion, it should not be done by default.

**Comment by tanjunyao7** [comment]
Hi Vladyslav,
you are right that SVD is expensive. But what I mean is that the users seldom need to construct a SE3/SO3 object from pure Eigen Matrix. Instead, the users frequently do manifold multiplication on SE3/SO3 to compute a pose from another poses. In the version 1.0, the poses multiplication are always followed by quaternion normalization https://github.com/strasdat/Sophus/blob/db218a249202fe63ac13248b5f565b0d385f6640/sophus/so3.hpp#L212-L229 
Since this nonmanifold-to-manifold-projection operation is done by default in frequently used group multiplication, I don't understand why this is not done by default in seldom used SO3 construction. If SVD is expensive, why not just normalize the result quaternion coefficients like the code above does?
